### PR TITLE
ECO-697 - client library

### PR DESCRIPTION
**Pull request by dwerner**
This PR separates the client into `client-lib` and `client`, moving the logic and responsibility for creating requests to the `casper-node` into `client-lib`.

### CODE REVIEW

**Review by Fraser999 — COMMENTED** [review]
I think we should keep the structure as per the `node`, i.e. keep the client as a single crate, but with a `[[bin]]` target living in a subdirectory `client/app`, and leaving all the library code in `client/src`.

This is partly for consistency, but also because if we split into two crates, we'd really want _both_ crates to be named `casper-client`.

Other than that, I'm only seeing a few nitpicks around import ordering and the likes.  The approach generally looks good to me.

By the way, please note https://github.com/CasperLabs/casper-node/pull/376 which will cause a few conflicts with this.  In particular, I stupidly removed the error enum and changed the methods in rpc.rs to do a `process::exit` rather than return `Result`s.  I guess I'll leave that as is in my PR since it makes that version slightly more user-friendly, but you'll need to unwind those changes when you merge mine into this one.

**Comment by dwerner** [comment]
Will wait for #376 to merge so I can factor in the changes from that.

**Comment by dwerner** [comment]
@Fraser999 I've changed my approach a bit in the hopes of both cleaning up the directory structure and abstracting the generic RPC-related options away from the functions themselves, while preserving the `verbose` behaviour as far as possible, though I did opt to remove the `process::exit` calls from within the library part itself.

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:251]: The approach I ended up taking was to return the full `JsonRpc` here back to the caller.  (I know I stripped out the `Result`, but I agree that we should keep that here)

There are a couple of reasons.  One - the caller can decide whether or not they need to work with an actual JSON value, or whether they just need to print the full `JsonRpc`.  The second more important point is that the RPC ID might be required by the caller, and that's lost of we only provide the JSON value here.

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:253]: We should be printing this if `verbose` (see https://github.com/CasperLabs/casper-node/blob/master/client/src/rpc.rs#L64-L70)

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:286]: I think this should be removed as mentioned in a previous comment.  Most callers from our own binary will be printing the same output, so doing here too seems a bit over the top.

If you disagree strongly, I could be persuaded.  I can see the lack of symmetry of printing the request but not the response as being a negative.

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:297]: Same with this - I'd prefer to leave the printing to the caller.

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:307]: Same again.

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:264]: If we're printing the `rpc_req` above in verbose mode, there's no need to print it here again.

File: client/src/balance.rs
Comment (Fraser999) [inline client/src/balance.rs:88]: Let's import `casper_client::RpcCall`.

(This is applicable in a quite a few places I believe)

**Review by Fraser999 — COMMENTED** [review]
This is looking really good.  A few issues which will take a bit of effort - the biggest one I expect being the lack of doc comments.

File: client/Cargo.toml
Comment (Fraser999) [inline client/Cargo.toml:18]:
```suggestion
name = "casper-client"
```

**Review by Fraser999 — CHANGES_REQUESTED** [review]
I've had a deeper look at this PR and have created a PR to your branch which addresses some of the issues.

Still to be done:
* I think we should move more of the logic out of the binary and into the library.  Specifically I'm thinking of the logic around creating a deploy.  Really the binary shouldn't be doing any heavy lifting - we should just parse args and pass these through to lib functions which actually construct the deploy
* Similarly for keygen.  We should have a lib function which takes `output_dir`, `force` and `algorithm`
* In `RpcCall`, the following methods could use fuller documentation of the format of some of their args:
    * `key` in `get_item()`
    * `purse_uref` in `get_balance()`
    * `transfer()`
* `RpcCall::transfer()` has a todo remaining
* `RpcCall::transfer()` has `URef` params, whereas `get_balance()` takes a URef as a string.  While I appreciate it's a bit inefficient, I think it'd be better to be consistent and change `get_balance()` to take a `URef` rather than a `String` then convert it back to a `String` in the `GetBalanceParams`
* We could look to reduce the duplication in the binary by moving the construction of the `RpcCall` to a function in `common`

**Comment by dwerner** [comment]
@Fraser999 I've tried to address your points above in atomic commits, the only point I was less sure about was your first:
>     * I think we should move more of the logic out of the binary and into the library.  Specifically I'm thinking of the logic around creating a deploy.  Really the binary shouldn't be doing any heavy lifting - we should just parse args and pass these through to lib functions which actually construct the deploy
For this I took it to mean that anything we were doing around `Deploy`s in general, but I could only see the supported CLTypes/CLValues and their related parsing as _somewhat_ matching what you're saying there. Let me know if I've missed the mark here.

File: client/lib/keygen.rs
Comment (Fraser999) [inline client/lib/keygen.rs:28]:
```suggestion
    Ed25519,
```

File: client/lib/keygen.rs
Comment (Fraser999) [inline client/lib/keygen.rs:31]:
```suggestion
    Secp256k1,
```

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:62]: Adding a blank line here will make the summary in the generated docs more appropriate.  Otherwise the entire doc comment is included in the summary.

```suggestion
    /// Queries the node for an item at `key`, given a `path` and a `state_root_hash`.
    ///
    /// - `state_root_hash` - Hash representing a pointer to the current state from which a value
```

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:88]:
```suggestion
    /// Gets the balance from a purse.
    ///
    /// - `state_root_hash` - Hash representing a pointer to the current state from which a balance
```

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:91]: I think it might be clearer to specify that it's the URef which can be obtained with `get_item`?

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:101]:
```suggestion
    /// Transfers an amount between purses.
    ///
    /// - `amount` - Specifies the amount to be transferred.
```

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:103]:
```suggestion
    /// - `source_purse` - Source purse in the sender's account that this amount will be drawn from.
    ///   If it is `None`, it will default to the account's main purse.
```

File: client/lib/rpc.rs
Comment (Fraser999) [inline client/lib/rpc.rs:113]: I think these should be codified into an enum to ensure they're mutually exclusive.

**Review by Fraser999 — COMMENTED** [review]
> For this I took it to mean that anything we were doing around Deploys in general, but I could only see the supported CLTypes/CLValues and their related parsing as somewhat matching what you're saying there. Let me know if I've missed the mark here. (That change should be isolated to 11bfe68)

I was a bit woolly in my description - sorry!

I think what we have here is ok actually (although maybe moving what is printed inside `show_arg_examples` should be moved to the lib too since the other simple-args stuff has been moved there?)

Let's discuss some further potential changes tomorrow, but in the meantime there are just a few more nitpicks :)

**Review by Fraser999 — APPROVED** [review]

**Review by Fraser999 — CHANGES_REQUESTED** [review]
As per comments in https://github.com/dwerner/casper-node/pull/2, let's change the lib's API to use `&str` parameters instead of domain types.

**Comment by dwerner** [comment]
Adding @henrytill as a reviewer - please have a look at the client to ensure I've preserved #414 properly.

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:218]: Copy-paste error - we should document `maybe_output_path` here.

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:225]: I doubt we need this now?

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:194]: I think we should keep a consistent style: either have all parameters described in the form of `* param - description` or use full sentences.  I don't really mind which, but I think we should be consistent.  This applies to a few of the doc comments added in this commit.

File: client/lib/deploy.rs
Comment (Fraser999) [inline client/lib/deploy.rs:27]:
```suggestion
pub(crate) struct Transfer {}
```

File: client/lib/deploy.rs
Comment (Fraser999) [inline client/lib/deploy.rs:30]:
```suggestion
pub(crate) struct ListDeploys {}
```

File: client/lib/error.rs
Comment (Fraser999) [inline client/lib/error.rs:39]:
```suggestion
    /// Failed to parse a `U128`, `U256` or `U512` from a string.
    #[error("failed to parse as U128, U256, or U512: {0:?}")]
```

File: client/lib/error.rs
Comment (Fraser999) [inline client/lib/error.rs:86]:
```suggestion
    /// Invalid `CLValue`.
```

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:181]:
```suggestion
/// Puts a `Deploy` on a node.
```

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:211]:
```suggestion
/// Makes a `Deploy` and outputs it to a file or stdout.
```

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:253]: No need for a new line here, and "it" should be "It".  Same a few lines below too.

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:271]: This is incorrect.  Currently the `transfer` subcommand allows for the full range of payment args, not just standard payment.

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:286]:
```suggestion
/// Signs a `Deploy` file.
```

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:341]:
```suggestion
    /// Payment amount - all other `payment_*` arguments are ignored and the standard payment
    /// contract is used if this is set.
```

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:298]: The comments on most of the member variables of `DeployStrParams`, `PaymentStrParams` and `SessionStrParams` are somewhat lacking.  These comments will be used by lib authors to understand what should be passed here, so they need to be a lot more explicit I think.

Probably referring to the help output from running `cargo r -- put-deploy --help` for each arg will give a much better indication of how these comments should look.

File: client/lib/lib.rs
Comment (Fraser999) [inline client/lib/lib.rs:342]: Since we have various constructors for this struct, I think if we make all the member variables private we'll be making it much harder to misuse.  Same comment for the `SessionStrParams`.

File: client/lib/parsing.rs
Comment (Fraser999) [inline client/lib/parsing.rs:33]:
```suggestion
}

```

File: client/lib/parsing.rs
Comment (Fraser999) [inline client/lib/parsing.rs:34]: I'm not sure there's any point to these small modules any more - they're pretty much all just called from one place each.  I'd be inclined to just move the parsing to code for the simpler ones to the callsites.  If you prefer to keep them here, I'm fine with that, but let's just drop the mods now.  So change `parsing::timestamp::parse()` to just something like `parsing::timestamp()`, etc.

Also, the comments are out of date now too.  They don't provide the arg now, they just parse it.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:218]: Thanks, done.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:225]: Right, missed this, thanks.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:253]: Thanks, done.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:271]: Ok, thanks for clarifying that - I think I was confused by the presence of two `*amount` arguments - one being `--amount` and the other `--payment-amount`. I'll add the full set of payment args back here.

File: client/lib/parsing.rs
Comment (dwerner) [reply client/lib/parsing.rs:35]: Addressed this, thanks.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:342]: Got it, thanks.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:194]: Got it, thanks.

File: client/lib/lib.rs
Comment (dwerner) [reply client/lib/lib.rs:298]: Got it, thanks.

**Review by Fraser999 — COMMENTED** [review]
This is looking good now - thanks for persevering!

Unfortunately there are a few more merge conflicts to resolve, but once they're done we should be ready to merge this.

**Comment by dwerner** [comment]
I've merged master and tested each of the subcommands, which lead to finding some more errors in parsing code and error handling. This has been resolved now and I'm hopeful this can merge soon. @Fraser999 thanks for all the effort on reviews.

File: client/src/deploy/creation_common.rs
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:142]: For a session path, we also need to collect the session args.

File: client/src/deploy/creation_common.rs
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:190]: For a payment path, we also need the payment args.

File: client/src/deploy/creation_common.rs
Comment (Fraser999) [inline client/src/deploy/creation_common.rs:241]:
```suggestion
    const ARG_HELP: &str =
        "RFC3339-like formatted timestamp, e.g. '2018-02-16 00:31:37'. If not provided, the \
        current time will be used. Note that timestamp is UTC, not local. See \
        https://docs.rs/humantime/latest/humantime/fn.parse_rfc3339_weak.html for more \
        information.";
```

**Review by Fraser999 — CHANGES_REQUESTED** [review]
After some test runs, there are still a couple of issues as per commented inline.

Also, it seems like `--list-deploys` now returns the whole block, which defeats the purpose of the subcommand.  I think we should remove `list_deploys` from the lib since it's just calling get-block.  Let's call `get_block` from the bin and display just the deploys' hashes from the response.  Probably still worth displaying them as a JSON array to be consistent with our other outputs.

**Review by henrytill — APPROVED** [review]
:+1: Looking good here :+1:

**Comment by dwerner** [comment]
> After some test runs, there are still a couple of issues as per commented inline.
> 
> Also, it seems like `--list-deploys` now returns the whole block, which defeats the purpose of the subcommand. I think we should remove `list_deploys` from the lib since it's just calling get-block. Let's call `get_block` from the bin and display just the deploys' hashes from the response. Probably still worth displaying them as a JSON array to be consistent with our other outputs.

I ended up pulling list-deploys out of the lib, but keeping the translation struct ListDeploysResult for json-serialization/prettification.

**Comment by dwerner** [comment]
I have a remaining issue here with `simple_args_*` arguments, but it also happens on master: (single quotes are stripped from the **value** portion of the `name:cl_type='value'`, unless it's all wrapped in double-quotes (or appropriately escaped single-quotes). Not sure if that issue should block this, but wanted to raise it because I saw this during testing.

```
dan@zaphod ~/D/c/client (master)> cargo run -- put-deploy --secret-key ../resources/local/secret_keys/node-1.pem --chain-name casper-charlie-testnet-1 --payment-amount 1000 --session-arg name_01:bool='false' --session-path ../target/wasm32-unknown-unknown/release/standard_payment.wasm
   Compiling casper-types v0.6.0 (/home/dan/Development/casper-node/types)
   Compiling casper-execution-engine v0.7.0 (/home/dan/Development/casper-node/execution_engine)
   Compiling casper-node v1.5.0 (/home/dan/Development/casper-node/node)
   Compiling casper-client v1.5.0 (/home/dan/Development/casper-node/client)
    Finished dev [unoptimized + debuginfo] target(s) in 47.31s
     Running `/home/dan/Development/casper-node/target/debug/casper-client put-deploy --secret-key ../resources/local/secret_keys/node-1.pem --chain-name casper-charlie-testnet-1 --payment-amount 1000 --session-arg 'name_01:bool=false' --session-path ../target/wasm32-unknown-unknown/release/standard_payment.wasm`
thread 'main' panicked at 'value in simple arg should be surrounded by single quotes unless it's a null optional value', client/src/deploy/creation_common.rs:541:13
stack backtrace:...
```

@momipsl Unrelated to my comment, but `--timestamp` arguments are now parsing the rfc339 format (here)[https://docs.rs/humantime/2.0.1/humantime/fn.parse_rfc3339_weak.html].

**Review by momipsl — APPROVED** [review]
Very nice work here.  I have a comment that is a little tangential but might lead to a follow on story.  I note that there are the following crate dependencies:

```
casper-execution-engine = { version = "0.7.0", path = "../execution_engine" }
casper-node = { version = "1.5.0", path = "../node" }
```

It is worth exploring the depth of these dependencies and whether they can be dropped and thus the client streamlined to a degree.

**Review by Fraser999 — APPROVED** [review]
Looks good now thanks.

**Pull request merged**
